### PR TITLE
Sarah/dolfinx cuas

### PR DESCRIPTION
**Pull request by SarahRo**
Code restructured:
- in implementation: folders dolfinx_contact, demos, tests
- setup.py added

Demos and tests using custom assemblers

### CODE REVIEW

File: implementation/demos/compare_cuas_snes.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    # and the bottom (contact condition)
+    if threed:
+        fname = "sphere"
+        create_sphere_mesh(filename=f"{fname}.msh")
+        convert_mesh(fname, "tetra")
+        with dolfinx.io.XDMFFile(MPI.COMM_WORLD, f"{fname}.xdmf", "r") as xdmf:
+            mesh = xdmf.read_mesh(name="Grid")
+        # mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, 10, 10, 20)
+
+        # def top(x):
```
Comment (jorgensd) [inline implementation/demos/compare_cuas_snes.py:73]: Could we remove code that is commented. If we want to keep these configurations, it is better to add an argument to the argparser.

File: implementation/demos/compare_cuas_snes.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            return x[2] < 0.15
+
+    else:
+        fname = "disk"
+        create_disk_mesh(filename=f"{fname}.msh")
+        convert_mesh(fname, "triangle", prune_z=True)
+        with dolfinx.io.XDMFFile(MPI.COMM_WORLD, f"{fname}.xdmf", "r") as xdmf:
+            mesh = xdmf.read_mesh(name="Grid")
+        # mesh = dolfinx.UnitSquareMesh(MPI.COMM_WORLD, 30, 30)
+
```
Comment (jorgensd) [inline implementation/demos/compare_cuas_snes.py:91]: Same comment as above.

File: implementation/demos/compare_cuas_snes.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        tdim = mesh.topology.dim
+        top_facets = dolfinx.mesh.locate_entities_boundary(mesh, tdim - 1, top)
+        bottom_facets = dolfinx.mesh.locate_entities_boundary(
+            mesh, tdim - 1, bottom)
+        top_values = np.full(len(top_facets), top_value, dtype=np.int32)
+        bottom_values = np.full(
+            len(bottom_facets), bottom_value, dtype=np.int32)
+        indices = np.concatenate([top_facets, bottom_facets])
+        values = np.hstack([top_values, bottom_values])
+        facet_marker = dolfinx.MeshTags(mesh, tdim - 1, indices, values)
```
Comment (jorgensd) [inline implementation/demos/compare_cuas_snes.py:122]: There should be a sorting command here, otherwise this will break in parallel, see: https://jorgensd.github.io/dolfinx-tutorial/chapter2/hyperelasticity.html?highlight=argsort

File: implementation/demos/compare_nitsche_snes.py [hunk 4]
```diff
@@ -68,7 +68,7 @@
         convert_mesh(fname, "tetra")
         with dolfinx.io.XDMFFile(MPI.COMM_WORLD, f"{fname}.xdmf", "r") as xdmf:
             mesh = xdmf.read_mesh(name="Grid")
-        #mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, 10, 10, 20)
+        # mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, 10, 10, 20)
```
Comment (jorgensd) [inline implementation/demos/compare_nitsche_snes.py:71]: Same comment as in previous codes

File: implementation/demos/test_nitsche_rigid_surface_cuas.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        mesh.topology.create_connectivity(tdim - 1, tdim)
+        with dolfinx.io.XDMFFile(MPI.COMM_WORLD, f"{fname}_facets.xdmf", "r") as xdmf:
+            facet_marker = xdmf.read_meshtags(mesh, name="Grid")
+        top_value = 2
+        bottom_value = 1
+        surface_value = 8
+        surface_bottom = 7
+
+    else:
+        # fname = "twomeshes"
```
Comment (jorgensd) [inline implementation/demos/test_nitsche_rigid_surface_cuas.py:78]: Remove commented code

File: implementation/dolfinx_contact/create_contact_meshes.py [hunk 6]
```diff
@@ -6,9 +6,9 @@
 import warnings
 
 import gmsh
-from mpi4py import MPI
+# from mpi4py import MPI
```
Comment (jorgensd) [inline implementation/dolfinx_contact/create_contact_meshes.py:9]: Remove if unused

File: implementation/dolfinx_contact/create_contact_meshes.py [hunk 7]
```diff
@@ -70,12 +70,79 @@ def create_circle_plane_mesh(filename: str):
...
+    # Add 4 points on circle (clockwise, starting in top left)
+    c_points2 = [gmsh.model.occ.addPoint(center2[0] + 2 * r * np.cos(angle), center2[1]
+                                         + 2 * r * np.sin(angle), center2[2]) for angle in angles]
+    arcs2 = [gmsh.model.occ.addCircleArc(
+        c_points2[i - 1], c2, c_points2[i]) for i in range(len(c_points2))]
+    curve2 = gmsh.model.occ.addCurveLoop(arcs2)
+    gmsh.model.occ.synchronize()
+    surface2 = gmsh.model.occ.addPlaneSurface([curve, curve2])
+    # # Create box
+    # p0 = gmsh.model.occ.addPoint(0, 0, 0)
```
Comment (jorgensd) [inline implementation/dolfinx_contact/create_contact_meshes.py:102]: Remove if unused

File: implementation/dolfinx_contact/create_contact_meshes.py [hunk 8]
```diff
@@ -70,12 +70,79 @@ def create_circle_plane_mesh(filename: str):
...
+    gmsh.finalize()
+
+
 def create_sphere_plane_mesh(filename: str):
     center = [0.0, 0.0, 0.0]
     r = 0.3
     angle = np.pi / 8
-    L = 1
-    B = 1
+    # L = 1
```
Comment (jorgensd) [inline implementation/dolfinx_contact/create_contact_meshes.py:144]: Remove if unused

File: implementation/dolfinx_contact/snes_against_plane.py [hunk 9]
```diff
@@ -72,7 +69,7 @@ def _u_D(x):
     dirichlet_dofs = dolfinx.fem.locate_dofs_topological(
         V, tdim - 1, facet_marker.indices[facet_marker.values == top_value])
     bc = dolfinx.DirichletBC(u_D, dirichlet_dofs)
-    bcs = [bc]
+    # bcs = [bc]
```
Comment (jorgensd) [inline implementation/dolfinx_contact/snes_against_plane.py:72]: Remove if unused

File: implementation/tests/test_dolfinx_cuas.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Load mesh and create identifier functions for the top (Displacement condition)
+    # and the bottom (contact condition)
+    if dim == 3:
+        fname = "sphere"
+        with dolfinx.io.XDMFFile(MPI.COMM_WORLD, f"{fname}.xdmf", "r") as xdmf:
+            mesh = xdmf.read_mesh(name="Grid")
+        # mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, 10, 10, 20)
+
+        # def top(x):
```
Comment (jorgensd) [inline implementation/tests/test_dolfinx_cuas.py:43]: Same comment as above

File: implementation/tests/test_dolfinx_cuas.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        facet_marker = dolfinx.MeshTags(mesh, tdim - 1, indices, values)
+
+        V = dolfinx.VectorFunctionSpace(mesh, ("CG", 1))
+        u = dolfinx.Function(V)
+        v = ufl.TestFunction(V)
+        du = ufl.TrialFunction(V)
+
+        # # Dirichlet boundary conditions
+        # def _u_D(x):
+        #     values = np.zeros((mesh.geometry.dim, x.shape[1]))
```
Comment (jorgensd) [inline implementation/tests/test_dolfinx_cuas.py:95]: Remove if unused

File: implementation/tests/test_dolfinx_cuas.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+        u = dolfinx.Function(V)
+        v = ufl.TestFunction(V)
+        u.interpolate(_u_initial)
+        metadata = {}  # {"quadrature_degree": q_deg}
+        dx = ufl.Measure("dx", domain=mesh, metadata=metadata)
+        ds = ufl.Measure("ds", domain=mesh, metadata=metadata,
+                         subdomain_data=facet_marker)
+        a = ufl.inner(sigma(u), epsilon(v)) * dx
+        # L = ufl.inner(dolfinx.Constant(mesh, [0, ] * mesh.geometry.dim), v) * dx
```
Comment (jorgensd) [inline implementation/tests/test_dolfinx_cuas.py:144]: Remove if unused

File: implementation/tests/test_dolfinx_cuas.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+        u.interpolate(_u_initial)
+        L = dolfinx.fem.Form(F)
+        b = dolfinx.fem.create_vector(L)
+
+        # Normal assembly
+        b.zeroEntries()
+        dolfinx.fem.assemble_vector(b, L)
+        # # Apply boundary conditions to the rhs
+        # dolfinx.fem.apply_lifting(b, [self._a], [self.bcs])
```
Comment (jorgensd) [inline implementation/tests/test_dolfinx_cuas.py:159]: Remove

File: implementation/tests/test_dolfinx_cuas.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+kt = dolfinx_cuas.cpp.contact.Kernel
+it = dolfinx.cpp.fem.IntegralType
+compare_matrices = dolfinx_cuas.utils.compare_matrices
+
+
+@pytest.mark.parametrize("theta", [1, -1, 0])
+@pytest.mark.parametrize("gamma", [10, 100, 1000])
+@pytest.mark.parametrize("dim", [2, 3])
+@ pytest.mark.parametrize("gap", [0.0, 0.02, -0.01])
+def test_contact_kernel(theta, gamma, dim, gap):
```
Comment (jorgensd) [inline implementation/tests/test_dolfinx_cuas.py:25]: It would be great if you could give a short description of what this test checks, as it is quite elaborate.

**Review by jorgensd — CHANGES_REQUESTED** [review]
